x86_64 assembly for `flint_mpn_mul_X`

albinahlback opened the pull request:
I think this should be sort of optimal. Perhaps one could further optimize it by aligning the instructions in some way. This effectively performs mul with X addmuls. For ARM, I think one should use my strategy seen in the first commit, but with the `adcx` and `adox` instructions, this should be faster.

I haven't tested it yet, but it probably doesn't work as it is my first time writing something proper in assembly.

Requires Intel ADX instruction set for `adcx` and `adox`. I don't know how to set up the configuration such that we can detect whether the CPU supports these instructions.

albinahlback commented:
I think we can implement 7x7 multiplication without using additional registers. Perhaps it is possible for 8x8 multiplication as well.

I haven't thought of a strategy when we cannot fit everything into the normal registers.

fredrik-johansson commented:
This looks excellent. Seems like you should be able to do all the unbalanced cases without difficulty as well.

albinahlback commented:
I think the following:
- Inline for 1x1 and 2x2,
- Create codelet for NxN for 2 < N < 9 (because then the regular registers are enough, and everything is similar) for which we generate symbols for this, and
- Think of how to proceed for N >= 9 (we should probably use XMM or YMM registers).

albinahlback commented:
> This looks excellent. Seems like you should be able to do all the unbalanced cases without difficulty as well.

Thanks! Yes, but let's see how hard it is for me to write codelets first...

fredrik-johansson commented:
Timings:

          gmp6.3.0     fredrik    albin
        mul_basecase
    4x4    1.21e-08   7.21e-09   5.88e-09
    5x5    1.43e-08   1.16e-08   8.73e-09
    6x6    1.84e-08   1.74e-08   1.18e-08
    7x7    2.31e-08   2.28e-08   1.57e-08

fredrik-johansson commented:
In addition to ``mul_nxm``, you might be able to do assembly ``addmul_nxm`` for similar sizes, then build larger sizes out of big block (e.g. 5x5) muls and addmuls.

(The addmuls themselves would be interesting for use in ``fmpz_addmul`` without requiring temporary storage.)

From size 16 or so probably a correctly optimized Karatsuba should start to win.

albinahlback commented:
Would mulhigh be interesting for Arb? Feels like an optimized routine for that would be very beneficial.

fredrik-johansson commented:
Yes, in fact that's why I started looking at basecase multiplication in the first place. Theoretically basecase mulhigh should be 2x faster than mpn_mul_basecase and in practice maybe a bit less; it's when my prototype code started being 2.1x faster than GMP that I realised we can improve full multiplications as well ;-)

For mulhigh, I think it's best to focus on the nxn case. I think we want two versions: one that computes n+1 high limbs, and one that computes n+2 high limbs. The first is what is implemented in MPFR's mulders.c and gives n accurate limbs minus n ulp error. This is fine when you want maximum speed, but the second version would be more useful when you have n-limb inputs and want n output limbs with guaranteed (nearly) correct rounding, since you get nearly a full limb of guard bits.

Of course, sqr and sqrhigh will be needed as well. Here you compute the s = contribution of terms below the diagonal, then do (s<<1)+t where t is the sum of the squares on the diagonal. The last step should be doable efficiently (interleaving multiplies with the shifts or adds so that things are nicely pipelined) in asm.

fredrik-johansson commented:
BTW, just doing

```
void
flint_mpn_mul_14(mp_ptr r, mp_srcptr x, mp_srcptr y)
{
    mp_limb_t t[14], u[14];
    mp_limb_t cy;

    flint_mpn_mul_7(r, x, y);
    flint_mpn_mul_7(r + 14, x + 7, y + 7);
    flint_mpn_mul_7(t, x, y + 7);
    flint_mpn_mul_7(u, x + 7, y);

    cy = mpn_add_n(t, u, 14);
    cy += mpn_add_n(r + 7, t, 14);
    mpn_add_1(r + 21, r + 21, 7, cy);
}
```

is 10% faster than GMP, and 15% faster with inlined add_s14a14a14 instead of mpn_add_n calls.

albinahlback commented:
@fredrik-johansson Can I ask you to check which ones passes the tests against your PR?

fredrik-johansson commented:
```
       gmp        albin
8    2.82e-08    1.94e-08    (1.45x faster)
9    3.51e-08    2.5e-08     (1.40x faster)
10   4.22e-08    3.22e-08    (1.31x faster)
```

Very nice!

albinahlback commented:
> ```
>        gmp        albin
> 8    2.82e-08    1.94e-08    (1.45x faster)
> 9    3.51e-08    2.5e-08     (1.40x faster)
> 10   4.22e-08    3.22e-08    (1.31x faster)
> ```
> 
> Very nice!

Sorry, but are you implying that they pass the tests?

I think N = 11 is feasible as well by pushing rdi into some xmm register, but after that I have no clue.

fredrik-johansson commented:
> Sorry, but are you implying that they pass the tests?

Yes, your functions pass the test code.

> I think N = 11 is feasible as well by pushing rdi into some xmm register, but after that I have no clue.

I suppose for larger N you may just have to create some virtual registers on the stack and waste some cycles pushing/popping them.

albinahlback commented:
> Yes, your functions pass the test code.

Haha wow, that's great!

Btw, perhaps I need to push xmm registers to the stack before populating them (I don't know, but that sounds reasonable)... This needs to be fixed.

fredrik-johansson commented:
When loading the operand into XMM registers, can't you do this faster (several words at once) using AVX2 instructions?

albinahlback commented:
> When loading the operand into XMM registers, can't you do this faster (several words at once) using AVX2 instructions?

Yes, but then one has to load them back to the registers. We can shuffle between each register, but then that's another instruction. But I think I will just use the stack for now.

albinahlback commented:
Okay, we can test if ADX is available through the `__ADX__` preprocessor definition.

albinahlback commented:
I added a "code generation" file. I just did it because I was lazy, not because it minimal or readable in any sense.

Currently, the supported sizes are $N \times N$ for $N = 3, \ldots, 13$.

I added a test as well, which I suppose will be a duplicate of Fredrik's test.

I will wait merging until Fredrik can confirm that these are as fast as expected. I'm mostly interested in the case N = 13 as that is when we really have to start putting things into the stack. But N > 8 is also interesting.

albinahlback commented:
Everything passes. Had a hard time understanding why Mac struggled, but they always define symbols prefixed with underscores.

albinahlback commented:
On Skylake against `mpn_mul_basecase`:
```
n =  3:    1.41e-08    6.59e-09   2.140x
n =  4:    1.90e-08    9.83e-09   1.933x
n =  5:    2.54e-08    1.43e-08   1.776x
n =  6:    3.34e-08    2.04e-08   1.637x
n =  7:    4.34e-08    2.76e-08   1.572x
n =  8:    5.45e-08    3.49e-08   1.562x
n =  9:    6.78e-08    4.69e-08   1.446x
n = 10:    8.16e-08    6.04e-08   1.351x
n = 11:    9.88e-08    7.83e-08   1.262x
n = 12:    1.14e-07    9.31e-08   1.224x
n = 13:    1.31e-07    1.15e-07   1.139x
```
n = 13 isn't that great. We should compare Karutsuba against this one.

fredrik-johansson commented:
I only expect Karatsuba to win around n = 16 due to the extra additions. Between that and the size where everything fits in registers, I suspect it will be better to reduce to four block products (or maybe just two block products if you can split just one operand?). This is probably also a reasonable way to do some of the unbalanced products in order to keep the code size down.

How does 12x12 perform if you implement it as a 2x2 product of 6x6 products, for example?

fredrik-johansson commented:
BTW, ``n_randtest`` isn't good enough for the test code. You need to use the GMP random functions (or something equivalent) to be sure to catch carry propagation bugs.

albinahlback commented:
I added uneven cases, that is $m \times n$ for $m = 3, \ldots, 8$ and $n = 1, \ldots, m - 1$.

I also fixed a bug where the overflag could be set when using `adc $0, \rX`, so I added `test %al, %al` afterwards to reset the flags.

albinahlback commented:
> BTW, `n_randtest` isn't good enough for the test code. You need to use the GMP random functions (or something equivalent) to be sure to catch carry propagation bugs.

Yes, this was a good catch. Noticed the bug because of this. Thanks!

fredrik-johansson commented:
> I added uneven cases, that is m×n for m=3,…,8 and n=1,…,m−1.

I notice that your nx1 code does the whole multiply and then writes the output. Here it's maybe faster to interleave the ``mov``s with the arithmetic? (I didn't try timing it yet.)

albinahlback commented:
> > I added uneven cases, that is m×n for m=3,…,8 and n=1,…,m−1.
> 
> I notice that your nx1 code does the whole multiply and then writes the output. Here it's maybe faster to interleave the `mov`s with the arithmetic? (I didn't try timing it yet.)

Yes. That is for sure one optimization. (I think the same applies for nx2.)

fredrik-johansson commented:
Could you also add 1x1, 2x1 and 2x2? Even if the C code is fine here (though I'm not sure about 2x2), it would be more uniform.

albinahlback commented:
> Could you also add 1x1, 2x1 and 2x2? Even if the C code is fine here (though I'm not sure about 2x2), it would be more uniform.

Sure.

albinahlback commented:
On Skylake, comparing against `mpn_mul_basecase` and `mpn_mul_1`:
```
           1      2      3      4      5      6      7      8      9     10     11     12

m =  1:  1.56x

m =  2:  2.02x  1.53x

m =  3:  1.67x  2.38x  1.93x

m =  4:  1.92x  1.91x  1.95x  1.96x

m =  5:  1.60x  1.69x  1.81x  1.87x  1.93x

m =  6:  1.60x  1.82x  1.71x  1.78x  1.69x  1.79x

m =  7:  1.61x  1.59x  1.73x  1.81x  1.61x  1.80x  1.68x

m =  8:  1.32x  1.42x  1.52x  1.62x  1.62x  1.62x  1.63x  1.56x

m =  9:  1.34x  1.36x  1.44x  1.43x  1.45x  1.44x  1.54x  1.52x  1.57x

m = 10:  1.46x  1.35x  1.18x  1.23x  1.28x  1.38x  1.38x  1.47x  1.40x  1.42x

m = 11:  1.32x  0.93x  0.99x  1.05x  1.16x  1.16x  1.19x  1.25x  1.23x  1.24x  1.31x

m = 12:  1.33x  0.83x  0.94x  0.87x  1.30x  1.11x  1.15x  1.20x  1.24x  1.22x  1.22x  1.26x
```
Major drop can be seen at $m = 10$ with $n = 3, 4, 5$, and $m = 11, 12$ with $n = 2, \ldots, 7$.

Currently no special routine is implemented for $n = 2$ for bigger $m$. I think it could be worth implementing.

albinahlback commented:
Added special case for n = 2 via chains. I think n = 3 would benefit from this as well, but the performance increase would be sort of minimal.

Current speed increase through `flint_mpn_mul_basecase`:
```
            1       2       3       4       5       6       7       8

m =  1:   1.00x

m =  2:   1.12x   1.20x

m =  3:   2.00x   1.91x   1.87x

m =  4:   1.75x   1.63x   1.69x   1.76x

m =  5:   1.80x   1.51x   1.55x   1.61x   1.64x

m =  6:   1.81x   1.46x   1.53x   1.55x   1.55x   1.56x

m =  7:   1.74x   1.46x   1.51x   1.47x   1.54x   1.48x   1.46x

m =  8:   1.50x   1.39x   1.47x   1.47x   1.50x   1.48x   1.51x   1.51x

m =  9:   1.35x   1.32x   1.39x   1.41x   1.44x   1.44x   1.45x   1.50x

m = 10:   1.35x   1.38x   1.39x   1.42x   1.47x   1.44x   1.46x   1.49x

m = 11:   1.45x   1.36x   1.40x   1.40x   1.46x   1.47x   1.46x   1.50x

m = 12:   1.34x   1.35x   1.34x   1.35x   1.41x   1.40x   1.66x   1.46x

m = 13:   1.30x   1.34x   1.33x   1.36x   1.41x   1.40x   1.40x   1.44x

m = 14:   1.28x   1.33x   1.31x   1.33x   1.38x   1.37x   1.39x   1.42x

m = 15:   1.31x   1.37x   1.30x   1.30x   1.37x   1.38x   1.40x   1.43x

m = 16:   1.29x   1.34x   1.29x   1.32x   1.37x   1.37x   1.39x   1.43x
```

albinahlback commented:
Btw, I think most of my routines does not work with aliasing. I think the only one working is `flint_mpn_mul_M_1`, where aliasing should work for all types of aliasing.

albinahlback commented:
> Btw, I think most of my routines does not work with aliasing. I think the only one working is `flint_mpn_mul_M_1`, where aliasing should work for all types of aliasing.

I haven't thought of any routines that works with aliasing, but usually `restrict` implies faster routines. For squaring algorithms, however, I would really prefer if they are aliasing compatible.

fredrik-johansson commented:
* Why do you think aliasing is an issue? ``mpn_mul`` doesn't support it either. Or are you saying that ``mul(y, x, x)`` might not work? I don't see any reason why that would be broken.

* I don't think we want a separate ``flint_mpn_mul_basecase`` function; rather we should just do the table lookups directly in ``flint_mpn_mul`` and ``flint_mpn_mul_n`` without a level of indirection.

* I think it's better to pad the tables by one entry so that we can lookup ``mul_funcs[alen][blen]`` rather than ``mul_funcs[alen - 1][blen - 1]``.

Do you think the current assembly routines are good enough for the sizes that are covered? In that case, I would be happy to merge now and do further improvements in future PRs (including rebasing #1674 on top of this).

fredrik-johansson commented:
I suppose you want to put back in at least the 9x9, 10x10 etc. products before merging this.

albinahlback commented:
> * Why do you think aliasing is an issue? `mpn_mul` doesn't support it either. Or are you saying that `mul(y, x, x)` might not work? I don't see any reason why that would be broken.

Oh, then it is fine! I was thinking about `mul(x, x, y)`.

>  * I don't think we want a separate `flint_mpn_mul_basecase` function; rather we should just do the table lookups directly in `flint_mpn_mul` and `flint_mpn_mul_n` without a level of indirection.

>  * I think it's better to pad the tables by one entry so that we can lookup `mul_funcs[alen][blen]` rather than `mul_funcs[alen - 1][blen - 1]`.

Sure, this is just a sketch!

albinahlback commented:
> Do you think the current assembly routines are good enough for the sizes that are covered? In that case, I would be happy to merge now and do further improvements in future PRs (including rebasing #1674 on top of this).

Yes, I think they are good enough. The Nx3 could be improved for N > 8, but the speedup would be pretty minor. The Nx2 case was more important.

> I suppose you want to put back in at least the 9x9, 10x10 etc. products before merging this.

Hmm, I don't know. I would like to go through our strategy before I start including this. There may be better alternatives than what I originally implemented. They performed pretty well for MxN when M = N, but not for inequalities.

fredrik-johansson commented:
> Hmm, I don't know. I would like to go through our strategy before I start including this. There may be better alternatives than what I originally implemented. They performed pretty well for MxN when M = N, but not for inequalities.

M = N is the most important case though, so it's nice if we have a speedup there.

fredrik-johansson commented:
Some statistics for how many calls are made to ``flint_mpn_mul`` or ``flint_mpn_mul_n`` for each size ``n x m`` when running various computations, via #1674. The counts are for min(n, 16) x min(m, 16) (so 17x17 counts as 16x16 etc).

``build/examples/pi 10000000``

          0 
          0   25869 
          0  100505   44744 
          0  369908   24533   21578 
          0   26581   84010   82551   62632 
          0   46460   36476    5234   74125    1862 
          0       5      38      90      20    5776    5328 
          0       0     107     640    1307     663   10988   13972 
          0       0       0    1419   12150    1555   17511   40249   43011 
          0       0       0     758   13900       0       0       0    2275    3422 
          0       1       0       0   28607    4913       0       0       0   10512   11340 
          0       0       0       2       0       0       0       0       1       4    1611    1731 
          0       0       0       0       7       0       0       0       0       2      25     197     184 
          0       0       0       0       4      23       0       0       0       1      12     127     742     670 
          0       0       0       0       0      20      80       0       0       0       1      46     631    2713    2522 
          0       0       0       0       1       0     103     770    3269   13916   12885    1696     222    3978   16029  168150 

``build/examples/hilbert_matrix 120``

         68 
        103       5 
         45  135542 1267753 
          0       0       0   25605 
          0   53151   67726   91626 1664730 
          0       0       0       0       5   11572 
          0       0       0       0       1       0    6824 
          0       0       0       0       0       0       0   10785 
          0       0       0       0       0       0       0       0   17081 
          0       0       0       0       0       0    1092    5616   19986 1817193 
          0       0       0       0       0       0       0       0       0     152    3339 


``build/examples/factor_integer 1000000000000000000000000000000000000000000000000000000000000019``

        439 
          0     219 
          0       0       0 
          0   17361     716 29087861 
          0   16219     150      30       0 
          0    3228     265     136       0       0 
          0   16235     349    7386       0       0    2297 
          0    2609     139      21       0       0       0      55 
          0   15336     187     172       0       0       5       0       0 
          0   13387     379    4138       0       0    3162       0       0       0 
          0    2640     203      30       0       0       4       0       0       0       0 
          0   11765     166     161       0       0       4       0       0       0       0       0 
          0    1563     246    1864       0       0    1718       0       0       0       0       0       0 
          0    9447     270    1851       0       0    1010       0       0       0       0       0       0       0 
          0    7474     213      54       0       0       7       0       0       0       0       0       0       0       0 
          0   46318    3100   36970       0       0   30387       0       0       0       0       0       0       0       0      54 


``build/examples/huge_expr``

       7336 
        539 2041922 
         23  174717 1615084 
         26   34374  223141  662833 
        108   23560   15187   64466  930987 
          0   24560   17008    7594   14731   89927 
          0   26946   17644    7040    7867   44031  279412 
          0   17554    9274    5514    3641    2982    2409  175556 
         10   12452   22272    4508    3166    2050    1779   39645  426854 
          3   15415    4455    4865    3586    2358    1691    2033    2170   98613 
          0   14403    3909    2733    4301    4349    1561    1856    1404    1414   35884 
          0   14186    3264    2110    2601    3178    1531    1610    1223     907     664   31966 
          1   16758    3021    1637    1557    1358    1995    1179    1060     656     294     518   39896 
          5   19939    3856    1131    1120     821    1079     900     813     282     182     154     154   39670 
          0   20677    7094     890    6474     811     831    1102     672     239     142     125     235    1298   22754 
         47  143404   69541   14943   40502    7610    3795   13610    8703    3246    1766    2970    1505    2113    5579  836314 

``build/examples/factor_polynomial "(1+x+2*y+3*z)^100-1"``

         12 
         14 12910004 
          2 1066214  521869 
          0  894163  600947  192350 
          0  736621  329930  108529     497 
          0  385816  153317    7081      65     279 
          0  200559   35031      22      64     174     714 
          0   77474      10       7       0       0       4      20 
          0 2654469 1974203 1176286  893358  580628  255882   29746 5070097 
         12 2153824 2000014 1603581 1091339  680489  479676  282679  107360 5077235 
          0       0       0       8       0       5       4       0       0       8      10 
          0       0      48      12      12       0       0       0       0       4       0       2 
          0       0       0       1       6       0       9       8       8       0       0       0  369816 
          0       1       0       0       0       0       0       0       0       4       0      11     129  647056 
          0       7       3       0       0       0       0       0       0       8       0       0      25      76  436678 
          0       5       0      11     203       1       0       8      14      47      40      59     103     143     299 2862441 


Some observations:

* The ``nx1`` counts are artificially low because we special-case by calling ``mpn_mul_1`` almost everywhere. We can also see that ``nx2``, ``nx3`` and ``nx4``, ... are quite common in some applications.
* In numerical applications, ``n x n`` and ``n x (n-1)`` are both common and deserve equal optimization.
* The last example would clearly benefit from all ``10 x m`` products being optimized.

albinahlback commented:
Will merge if it passes all the tests. `mpn_extras/mul_basecase.c` has to then be merged with `mpn_extras/mul.c`.
